confgenerator/fluentbit/modular.go

```diff
@@ -24,22 +24,24 @@ import (
 type Component struct {
 	// Kind is "INPUT", "FILTER", "PARSER", etc.
 	Kind string
-	// Config is a set of key-value configuration pairs
-	Config map[string]string
+	// Config is a slice of slices containing key & value
+	// This allows duplicate keys and specified order
+	Config [][2]string
 }
 
 func (c Component) generateSection() string {
 	var lines []string
 	var maxLen int
-	for k := range c.Config {
-		if len(k) > maxLen {
-			maxLen = len(k)
+	for _, line := range c.Config {
+		if len(line[0]) > maxLen {
+			maxLen = len(line[0])
 		}
 	}
-	for k, v := range c.Config {
-		lines = append(lines, fmt.Sprintf("    %-*s %s", maxLen, k, v))
+
+	for _, line := range c.Config {
+		lines = append(lines, fmt.Sprintf("    %-*s %s", maxLen, line[0], line[1]))
 	}
-	sort.Strings(lines)
```

Comment:
I vaguely recall we added explicit sorting to ensure the generated config order is deterministic. Is there a reason that prevents us from sorting them in the new structure? If there are repeated keys, we can sort by key first, then value I guess.

Are the field order difference in the `golden_fluent_bit_parser.conf` files below are related to this?

Reply:
Yes, the field order differences in the golden configs are due to this. The issue with sorting in the new structure is some of the repeated keys require intentional ordering. Rule for multiline might be more flexible, I haven't tested it, but Parser for multi-parser filter definitely uses ordering. We can name them so they always sort correctly, but something mentioned in the other ticket specified "Grep" parsers requiring specifically ordered Exclude and Regex fields, which I'm not as familiar with but might be harder to guarantee ordering correctly.

Reply:
I think my personal inclination is that it's better overall to just keep two separate fields (a sorted map for almost everything, and RepeatedConfig as `[][2]string` for the very limited cases where we need to control the order of options). I don't think the benefits of the struct simplification outweigh the more cumbersome config construction and the out-of-order generated configs.

Reply:
Sure, I can get on board with that -- it keeps the diff simpler too. One small nit, I'd rename `RepeatedConfig` to `OrderedConfig`.

Reply:
Oh, gotcha. I didn't see that `RepeatedConfig` was currently still a map.

(BTW, it's definitely a gross hack, but I had originally planned that if we ever needed repeated config in an edge cases it could be "handled" by just putting newlines in the value of a key... I think it's probably far better to build an explicit marshaling interface for it, though.)